gentypes.go

```diff
@@ -0,0 +1,211 @@
+	switch *tp {
+	case pb.Type_BOOL:
+		return "bool", nil
+	case pb.Type_ANY:
+		return "interface{}", nil
+	case pb.Type_INT:
+		return "int", nil
+	case pb.Type_STRING:
+		return "string", nil
+	case pb.Type_EMPTY:
+		return "nil", nil
+	case pb.Type_FLOAT, pb.Type_DECIMAL:
+		return "float64", nil
+	case pb.Type_BYTES:
+		return "[]byte", nil
+	case pb.Type_DATE, pb.Type_DATETIME:
+		return "time.Time", nil
+	}
+	// Type_XML, Type_UUID
+	return "", fmt.Errorf("unsupported type primitive %s", tp.String())
+}
+
+// GenSimpleType returns Golang type string for non-composite types (no lists and sets)
+func GenSimpleType(t *pb.Type) (string, error) {
+	if pType := t.GetPrimitive(); pType != pb.Type_NO_Primitive {
+		return GenPrimitiveType(&pType)
+	}
+	if t.GetTypeRef() != nil {
+		path := t.GetTypeRef().GetRef().GetPath()
+		if len(path) != 1 {
+			return "", fmt.Errorf("cannot handle type reference with more than one path")
+		}
+		str := path[0]
+		if !strings.HasPrefix(str, "map of") {
+			return str, nil
+		}
+		str = strings.TrimPrefix(str, "map of")
+		str = strings.Trim(str, " ")
+		m := strings.Split(str, ":")
+		if len(m) != 2 {
+			shouldStr := `should be map of KeyType:ValueType`
+			return "", fmt.Errorf("bad map definition '%s' (%s)", str, shouldStr)
+		}
+		return fmt.Sprintf("map[%s]%s", m[0], m[1]), nil
+	}
+	return "", fmt.Errorf("type %v is neither primitive nor reference", t)
+
+}
+
+// GetLine returns the line for a given Sysl type from its SourceContext
+func GetLine(t *pb.Type) (int32, error) {
+	if t.GetPrimitive() != pb.Type_NO_Primitive || t.GetTypeRef() != nil {
+		return t.GetSourceContext().Start.Line, nil
+	} else if t.GetList() != nil {
+		return t.GetList().GetType().GetSourceContext().Start.Line, nil
+	} else if t.GetSet() != nil {
+		return t.GetSet().GetSourceContext().Start.Line, nil
+	} else if t.GetTuple() != nil {
+		for _, t2 := range t.GetTuple().GetAttrDefs() {
+			return GetLine(t2)
+		}
+	}
+	return 0, fmt.Errorf("unknown type %v for getting line", t)
+}
+
+//NamesSortedBySourceContext sorts the keys of the input types according to occurrence
+// in Sysl definition file, derived from SourceContext
+func NamesSortedBySourceContext(types map[string]*pb.Type) ([]string, error) {
+	type lineName struct {
+		name string
+		line int32
+	}
+	size := len(types)
+	lineNames := make([]lineName, size)
+	i := 0
+	for name, t := range types {
+		line, err := GetLine(t)
+		if err != nil {
+			return nil, err
+		}
+		lineNames[i] = lineName{name, line}
+		i++
+	}
+	sort.Slice(lineNames, func(i, j int) bool {
+		return lineNames[i].line < lineNames[j].line
+	})
+	result := make([]string, size)
+	for i := 0; i < size; i++ {
+		result[i] = lineNames[i].name
+	}
+	return result, nil
+}
+
+// SplitUppercase splits into fields at all upper case letters and converts
+// fields to lower case
+func SplitUppercase(str string) []string {
+	result := []string{}
+	var i int
+	for s := str; s != ""; s = s[i:] {
```

Comment:
Is string slicing O(1) or O(n)?

Reply:
> Since strings are immutable
So I guess it's O(n) (or even more), but n is hopefully pretty small - we are talking about strings holding variable names after all.

Reply:
I would run into problems with the indexFunc though. I guess I'd just have to rewrite it to find All uppercase letters in one wash. Again - I'm not sure if this is worth the effort when usually hopefully dealing with pretty short strings?

Reply:
Okokok - I was pretty happy with copying this from the net, but alas: 
I have re-written it so now it's O(n) with n being the string length. (can't be any better because of the copying of the string parts I believe.)